File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {       
        int n = candidates.size();
        vector<vector<int>> ans;
        vector<int> sol;

        function<void(int, int)> backtracking = [&](int i, int sum) {

            if(sum == target){
                ans.push_back(sol);
                return;
            }
            if(sum>target || i==n)
                return;

            backtracking(i+1, sum);
            sol.push_back(candidates[i]);
            backtracking(i,sum + candidates[i]);
            sol.pop_back();
        };
        backtracking(0,0);
        return ans;
    }
};

File: 0039-combination-sum/0039-combination-sum.kt
class Solution {
    fun combinationSum(nums: IntArray, target: Int): List<List<Int>> {
        val ans = mutableListOf<List<Int>>()
        val sol = mutableListOf<Int>()
        val n = nums.size

        fun backtracking(i : Int, sum : Int){
            if(sum == target){
                ans.add(ArrayList(sol))
                return
            }
            if(sum> target || i == n) return

            backtracking(i+1,sum)
            sol.add(nums[i])
            backtracking(i,sum+nums[i])
            sol.removeAt(sol.size-1)
        }
        backtracking(0,0)
        return ans
    }
}